### ISSUE TITLE
"RangeError: Maximum call stack size exceeded" when mocking a large library

### ISSUE THREAD

**Issue by joshuavial** [Typescript-TDD/ts-auto-mock#83]
https://github.com/joshuavial/ts-auto-mock/commit/1647fc0afa825a02eb6984f2d477b05810310188 has an example of ts-auto-mock failing to parse a fairly large definition file from mongoose (https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/mongoose/index.d.ts)

```
Child process failed to process the request: RangeError: Maximum call stack size exceeded
    at getTypeOfNode (/ts-auto-mock/node_modules/typescript/lib/typescript.js:60536:20)
    at Object.getTypeAtLocation (/ts-auto-mock/node_modules/typescript/lib/typescript.js:32265:31)
    at /ts-auto-mock/dist/transformer/index.js:1:6913
```

I'm wondering if it's a problem with the size of the definition causing the GetDeclaration recursion to overrun the stack size. Hopefully there is a simple work around which doesn't involve a lot of re-architecture, but happy to lend a hand if that's what it takes to parse the library.

**Comment by uittorio** [comment]
The reason of this issue could be a recursive interface that ts-auto-mock is not handling correctly. I'll try to isolate the issue and let you know. Thank for the feedback

**Comment by uittorio** [comment]
I've some investigation and there are plenty of unsopported features. To support that interface it will take a while and I will have to create specific bugs. I reccomend isolate the usage of Mongose to a specific functionality so you will not use this library to mock that interface.

**Comment by joshuavial** [comment]
Thanks for looking in to it @uittorio - happy to help chip away at the backlog when you create it as I'd love to be able to point ts-auto-mock at arbitrary libraries. Of all the ts mocking libraries out there ts-auto-mock is the one that I most want to exist :)

**Comment by uittorio** [comment]
I've found the main cause of the issue. when a type extends another type that has a property with this it will fail

```ts 
interface TestB {
  property: this
}
interface Test extends TestB {}
```

I have in mind few differet solutions. I'll create an issue and I'll work on it this weekend

**Comment by uittorio** [comment]
this issue is related to #85

**Comment by uittorio** [comment]
@joshuavial  
Hi, I've finished the investigation related to the issue when trying to mock mongoose.
I've fixed one of them #85. It was a long one, but thank you very much for spending time writing the bug. It allows ts auto mock to be extended in a powerful way.

Mostly of the issue reference above are not blocking the usage of the library. The only one that is blocking it is Recursion in Intersection #90. After that one is completed you can mock that library, however few functionality will not be there. I honestly will be surprised that you need the entire Interface mocked 100%. 
For example typeof will be null

**Comment by uittorio** [comment]
@joshuavial A new version has been released 1.3.1 that doesn't crash when using that interface. Few types will not work. Please read the documentation for more details.

**Comment by uittorio** [comment]
I'll close this bug for inactivity :). Feel free to re open it if you find the same issue
